lib/logthrdest/logthrdestdrv.c

```diff
@@ -966,6 +966,8 @@ _construct_worker(LogThreadedDestDriver *self, gint worker_index)
        * single worker we have and all Worker related state is in the
        * (derived) Driver class. */
 
+      main_loop_threaded_worker_clear(&self->worker.instance.thread);
+      main_loop_threaded_worker_init(&self->worker.instance.thread, MLW_THREADED_OUTPUT_WORKER, &self->worker.instance);
```

Comment:
I think it would be better to use the init_instance function of LogThreadedDestWorker here instead of in the constructor and similarly call log_threaded_dest_worker_free_method()  in the deinit path.

e.g. something like this untested patch:

```
$ git diff .
diff --git a/lib/logthrdest/logthrdestdrv.c b/lib/logthrdest/logthrdestdrv.c
index b07f79cc1..fb8f361ef 100644
--- a/lib/logthrdest/logthrdestdrv.c
+++ b/lib/logthrdest/logthrdestdrv.c
@@ -966,12 +966,13 @@ _construct_worker(LogThreadedDestDriver *self, gint worker_index)
        * single worker we have and all Worker related state is in the
        * (derived) Driver class. */
 
+      log_threaded_dest_worker_init_instance(&self->worker.instance, self, 0);
+      _init_worker_compat_layer(&self->worker.instance);
       return &self->worker.instance;
     }
   return self->worker.construct(self, worker_index);
 }
 
-
 void
 log_threaded_dest_driver_set_max_retries_on_error(LogDriver *s, gint max_retries)
 {
@@ -1258,6 +1259,8 @@ log_threaded_dest_driver_deinit_method(LogPipe *s)
       for (int i = 0; i < self->created_workers; i++)
         log_threaded_dest_worker_free(self->workers[i]);
     }
+  else
+    log_threaded_dest_worker_free_method(&self->worker.instance);
 
   return log_dest_driver_deinit_method(s);
 }
@@ -1268,7 +1271,6 @@ log_threaded_dest_driver_free(LogPipe *s)
 {
   LogThreadedDestDriver *self = (LogThreadedDestDriver *)s;
 
-  log_threaded_dest_worker_free_method(&self->worker.instance);
   g_mutex_clear(&self->lock);
   g_free(self->workers);
   log_dest_driver_free((LogPipe *)self);
@@ -1294,6 +1296,4 @@ log_threaded_dest_driver_init_instance(LogThreadedDestDriver *self, GlobalConfig
   self->retries_on_error_max = MAX_RETRIES_ON_ERROR_DEFAULT;
   self->retries_max = MAX_RETRIES_BEFORE_SUSPEND_DEFAULT;
   g_mutex_init(&self->lock);
-  log_threaded_dest_worker_init_instance(&self->worker.instance, self, 0);
-  _init_worker_compat_layer(&self->worker.instance);
 }
```

Reply:
@bazsi I can't completely reconstruct the worker in compat mode, as the worker does not have a constructor method that I can call (it is bound to the driver in compat mode, the driver's and the worker's lifetimes must be the same).

If I call init_instance on the path you suggested, it will just undo everything that the child class has overridden in its constructor (virtual table, etc.).

My understanding is that if I tried to fix the reload issue cleanly and correctly in compat mode, I would just end up reimplementing the `worker.construct` method, which would be just the equivalent of removing the compat mode.
I didn't want to get rid of the compat mode, so I did this half-reinitialization trick.

Reply:
I wrote my previous comment before receiving https://github.com/syslog-ng/syslog-ng/pull/4588#discussion_r1294590333.

I'll soon check and understand that patch too.

Reply:
Ok, this seems to be working after I fixed the above patch. `_destroy_workers` didn't work properly.

Reply:
FYI the Python destination used `worker.instance` before calling the parent's init() method.

Reply:
I see. I missed that. Sorry for the noise.

Reply:
I was wrong, sorry. Everything is fine now.